[20446] Add error handling to form labels

oliverguenther opened the pull request:
This re-enables an `-error` state for labels when the associated
field is erroneous.

Also shows an error icon next to the label and changes the font style.

![bildschirmfoto 2015-11-30 um 13 33 38](https://cloud.githubusercontent.com/assets/459462/11471575/fc724e62-9766-11e5-9efd-3b81a95c314b.png)

Relevant work package: https://community.openproject.org/work_packages/20446

oliverguenther committed:
Add error handling to form labels

This re-enables an `-error` state for labels when the associated
field is erroneous.

Also shows an error icon next to the label and changes the font style.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-fix-20446-error-labels.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://oliverguenther-fix-20446-error-labels.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

oliverguenther commented:
There may be inconsistencies when, e.g., plugins do not use the form helper to generate fields, but produce labels on their own. I've fixed this for the reported case (costs), but there may be more. The resulting error is not critical, but simply a visual bug that those labels will not have the adapted style.

Maybe @lindenthal can comment on whether the error labels are something we should advance with. I believe its the most UX-friendly in regards to seeing, but visually impaired users.

ulferts commented:
I like the improvement both codewise as well as usabilitywise. And I'd still put this on hold because if we proceed this way, it should be done throughout the application. Right now, even without the plugins, the work package form behaves differently from the rest. But what @oliverguenther did should give @lindenthal a good sample to assess if the approach taken fit's the application's needs.

On top of that, I have one minor remark regarding the accessibility. The color contrast is 4.89:1 (`#CA3F3F` on `#FFF`) which passes the AA requirements for fonts below 18px of 4.5:1 defined by [WCAG](http://www.w3.org/TR/WCAG20/) just barely. It fails the AAA requirements for fonts below 18px of 7:1 by a wide margin. So I think that another color would be beneficial.

furinvader committed:
Remove unnecessary type declarations in routing config

furinvader committed:
Add ng-annotate loader for typescript files

furinvader committed:
Add openproject.services to openproject.wpButtons module dependencies

ulferts committed:
establish accessibility for work package attributes

furinvader committed:
Implement wp-buttons base module

Delivering two base controller classes and a directive helper.

furinvader committed:
Add WP button template file

So the module may test a generic button directive.

ulferts committed:
writing custom directive for accessible-wp-attribute

oliverguenther committed:
Fix missing label for password confirmation

furinvader committed:
Move WP accessible attribute directive to components

furinvader committed:
Abbreviate name of the WP accessible attribute directive

furinvader committed:
Add some minor changes

Using $digest is fine, although $apply is the recommended way to go.

crohr committed:
Fix docker nginx config for serving fonts

furinvader committed:
Merge pull request #4075 from furinvader/features/button_module

Implement base module for WP button components

furinvader committed:
Implement keepTab service

Intendet to replace the latestTab controller dependency.

furinvader committed:
Implement WP details view button directive

furinvader committed:
Implement WP show view button directive

furinvader committed:
Get rid of latestTab

openprojectci committed:
Update reference to OpenProject-Translations

HDinger committed:
Set default color of sidebar menu for all color schemes

oliverguenther committed:
Merge pull request #4077 from oliverguenther/fix/22490-password-reset-confirmation

Fix missing label for password confirmation

oliverguenther committed:
Merge pull request #4080 from HDinger/fix/22433-dark-theme-sort-by-fields-are-almost-not-visible

[22433] Dark Theme: "Sort By" fields are almost not visible

HDinger committed:
Replace ui-select with normal select box for better accessibility

furinvader committed:
Fix WP list controller test

oliverguenther committed:
Merge pull request #4081 from HDinger/fix/22315-replace-UI-select-on-timelines

[22315] Replace UI select on timelines

furinvader committed:
Make inplaceEditAll.state a getter-only property

oliverguenther committed:
Merge pull request #4079 from furinvader/refactoring/show_and_details_view_buttons

Refactor show and details view button

oliverguenther committed:
Merge pull request #4083 from furinvader/refactoring/routing-config

Remove unnecessary type declarations in routing config

furinvader committed:
Merge pull request #4084 from furinvader/feature/ng-annotate_typescript

Add ng-annotate loader for typescript files

oliverguenther committed:
Merge pull request #4085 from furinvader/refactoring/inplace-edit-all

Make inplaceEditAll.state a getter-only property

HDinger committed:
Introduced required stars for table headers in CF Activities

openprojectci committed:
Update reference to OpenProject-Translations

openprojectci committed:
Update reference to OpenProject-Translations

ulferts committed:
Merge pull request #6 from furinvader/ulferts/feature/accessible_wp_attributes

Move directive definition away from legacy file structure

ulferts committed:
adapt test

oliverguenther committed:
Fix build-up of class attributes

Due to an error in the parameter handling of `MenuManager::MenuItem`,
every request lead to an additional three class attributes in plugin
project menu items.

After some extensive debugging with @ulferts, we found this to be the definite cause of the crashing Mobile Safari.

oliverguenther committed:
Merge pull request #4078 from ulferts/feature/accessible_wp_attributes

Feature/accessible wp attributes

openprojectci committed:
Update reference to OpenProject-Translations

openprojectci committed:
Update reference to OpenProject-Translations

openprojectci committed:
Update reference to OpenProject-Translations

ulferts committed:
Merge pull request #4090 from oliverguenther/fix/22378-fix-class-attributes

[22378] Fix build-up of class attributes

ulferts committed:
Merge pull request #4086 from HDinger/fix/22375-star-icon-for-some-required–custom-fields-missing

[22375] Star icon for required custom fields of type "Activity (time tracking)" not displayed

oliverguenther committed:
Create error color with higher contrast on whites

Contrast on #FFF: 7.5:1

oliverguenther committed:
Add separate span with error message for the current field

oliverguenther committed:
Refactor required star into separate spans

Seperate spans for visible and impaired users allows us to replace the
spelled out Asterisk from pseudo content with a proper aria-label.
